lib/index.js

```diff
@@ -36,143 +68,112 @@ module.exports = {
-          cb();
-        });
-        server.on('uncaughtException', function(req, res, route, err) {
-          logger(req, res, route, err);
-          if (!res.headersSent) {
-            res.send(err.statusCode || 500, { type: err.type || 'Error', message: err.message });
-          }
+      });
+      routes[api.basePath] = [{
+        name: 'index',
+        method: 'get',
+        version: api.info.version,
+        handler: indexRouteHandler(api)
+      }];
+    });
+
+
+    const corsMap = {};
+    keys(routes).sort().reverse().forEach((routePath) => {
+      corsMap[routePath] = { shouldSet: true, expectedHeaders: new Set(), allowedMethods: new Set(['OPTIONS']) };
+      routes[routePath].forEach((action) => {
+        if (action.method === 'opts') { corsMap[routePath].shouldSet = false; }
+        corsMap[routePath].allowedMethods.add(action.method.toUpperCase());
+        (action.expectedHeaders || []).forEach((header) => {
+          corsMap[routePath].expectedHeaders.add(header);
         });
 
-        server.listen(
-          config.get('port'),
-          config.get('host'),
-          function() {
-            debug('%s listening at %s', server.name, server.url);
-          }
+        debug('routing %s@%s for %s', routePath, action.version, action.name);
+        server[action.method](
+          { path: routePath, version: action.version },
+          cors.headers(),
+          action.handler
         );
+      });
+    });
 
-        return server;
+    keys(corsMap).forEach((route) => {
+      const opts = cors.preflight(corsMap[route]);
+      if (opts) { server.opts(route, opts); }
+    });
 
-      }).catch(function(err) {
-        debug(err);
-        throw err;
-      });
+    server.on('after', function(req, res, route, err) {
+      logger(req, res, route, err);
+    });
+    server.on('NotFound', function(req, res, err, cb) {
+      res.send(404, { type: 'NotFound', message: err.message });
+      cb();
+    });
+    server.on('MethodNotAllowed', function(req, res, err, cb) {
+      res.send(405, { type: 'MethodNotAllowed', message: err.message });
+      cb();
+    });
+    server.on('VersionNotAllowed', function(req, res, err, cb) {
+      res.send(400, { type: 'VersionNotAllowed', message: err.message });
+      cb();
+    });
+    server.on('InternalServer', function(req, res, err, cb) {
+      res.send(500, { type: 'ServerError', message: err.message });
+      cb();
+    });
+    server.on('uncaughtException', function(req, res, route, err) {
+      logger(req, res, route, err);
+      if (!res.headersSent) {
+        res.send(err.statusCode || 500, { type: err.type || 'Error', message: err.message });
+      }
+    });
+
+    const listenDefer = defer();
+    server.listen(config.get('port'), config.get('host'), () => {
+      debug('%s listening at %s', server.name, server.url);
+      listenDefer.resolve();
+    });
 
+    await listenDefer.promise;
+    return server;
   },
 
   stop: function() {
```

Comment:
I think close has a callback too, should probably do the same defer/resolve promise thing here?